hazelcast/src/main/java/com/hazelcast/quorum/QuorumServiceImpl.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.quorum;
+
+import com.hazelcast.core.Member;
+import com.hazelcast.core.MembershipEvent;
+import com.hazelcast.instance.Node;
+import com.hazelcast.spi.MutatingOperation;
+import com.hazelcast.spi.ReadonlyOperation;
+import com.hazelcast.nio.ClassLoaderUtil;
+import com.hazelcast.spi.EventPublishingService;
+import com.hazelcast.spi.EventService;
+import com.hazelcast.spi.MemberAttributeServiceEvent;
+import com.hazelcast.spi.MembershipAwareService;
+import com.hazelcast.spi.MembershipServiceEvent;
+import com.hazelcast.spi.NamedOperation;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.QuorumAwareService;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.util.ConstructorFunction;
+import com.hazelcast.util.ExceptionUtil;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized;
+
+
+/**
+ * Service containing logic for cluster quorum.
+ */
+public class QuorumServiceImpl implements EventPublishingService<QuorumEvent, QuorumListener>, MembershipAwareService,
```

Comment:
I think relying on the cluster to determine quorum is not the way forward. Every member can keep track of the members list by registering a member listener. There is no need to exchange with other members that this member can or can not reach quorum. Everyone will come to the same result. 

This is an architectural issue we need to address.